ApplicationLibCode/ReservoirDataModel/RigMswCenterLineCalculator.cpp

```diff
@@ -0,0 +1,331 @@
+            {
+                const auto& [gridIndex, cellIndex] = gridAndCellIndices.front();
+
+                const RigCell& cell       = eclipseCaseData->grid( gridIndex )->cell( cellIndex );
+                cvf::Vec3d     whStartPos = cell.faceCenter( cvf::StructGridInterface::NEG_K );
+
+                // Add extra coordinate between cell face and cell center
+                // to make sure the well pipe terminated in a segment parallel to z-axis
+
+                cvf::Vec3d whIntermediate = whStartPos;
+                whIntermediate.z()        = ( whStartPos.z() + cell.center().z() ) / 2.0;
+
+                RigWellResultPoint resPoint;
+                for ( const auto& resBranch : resultBranches )
+                {
+                    for ( const auto& respoint : resBranch.m_branchResultPoints )
+                    {
+                        if ( respoint.segmentId() == firstSegment )
+                        {
+                            resPoint = respoint;
+                            break;
+                        }
+                    }
+                }
+
+                cellCenterCoords.push_back( whStartPos );
+                cellCenterResultPoints.push_back( resPoint );
+                cellCenterCoords.push_back( whIntermediate );
+                cellCenterResultPoints.push_back( resPoint );
+            }
+        }
+
+        for ( const auto& [segmentId, gridAndCellIndices] : wellBranch.m_segmentsWithGridCells )
+        {
+            for ( const auto& [gridIndex, cellIndex] : gridAndCellIndices )
+            {
+                const RigCell& cell = eclipseCaseData->grid( gridIndex )->cell( cellIndex );
+                cvf::Vec3d     pos  = cell.center();
+
+                RigWellResultPoint resPoint;
+
+                // The result point is only used to transport the grid index and cell index
+                // The current implementation will propagate the cell open state to the well segment from one cell to
+                // the next. It this is misleading, the two following lines can be removed.
+                resPoint.setGridIndex( gridIndex );
+                resPoint.setGridCellIndex( cellIndex );
+
+                cellCenterCoords.push_back( pos );
+                cellCenterResultPoints.push_back( resPoint );
+            }
+        }
+
+        const auto simWellBranch = addCoordsAtCellFaceIntersectionsAndCreateBranch( cellCenterCoords, cellCenterResultPoints, eclipseCaseData );
+
+        simWellBranches.emplace_back( simWellBranch );
+    }
+
+    return simWellBranches;
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+SimulationWellCellBranch
+    RigMswCenterLineCalculator::addCoordsAtCellFaceIntersectionsAndCreateBranch( const std::vector<cvf::Vec3d>          branchCoords,
```

Comment:
Consider adding verification of `branchCoord.size() == resultPoints.size()`, to enusre valid `resPointIdx` in for-loop